File: 2021/python/day02.py
def part_1(commands):
    x = 0
    y = 0
    for direction, distance in commands:
        distance = int(distance)
        if direction == "forward":
            x += distance

        elif direction == "up":
            y -= distance

        elif direction == "down":
            y += distance
    return x * y


def part_2(commands):
    x = 0
    y = 0
    aim = 0
    for direction, distance in commands:
        distance = int(distance)
        if direction == "forward":
            x += distance
            y += aim * distance

        elif direction == "up":
            aim -= distance

        elif direction == "down":
            aim += distance
    return x * y


if __name__ == "__main__":
    with open("../input/day02.txt", "r") as commands:
        commands = [tuple(d.strip().split(" ")) for d in commands.readlines()]
        print("Part 1:", part_1(commands))
        print("Part 2:", part_2(commands))

File: 2021/elixir/lib/advent/day_02.ex
defmodule Advent.Day02 do
  def main(commands) do
    commands =
      commands
      |> Enum.map(fn raw ->
        [dir, dist] = String.split(raw)
        {String.to_atom(dir), String.to_integer(dist)}
      end)

    {commands |> part_1(), commands |> part_2()}
  end

  def part_1(commands) when is_list(commands) do
    commands
    |> Enum.reduce([x: 0, y: 0], &part_1/2)
    |> calculate_answer()
  end

  def part_1({:forward, dist}, x: x, y: y) do
    [x: x + dist, y: y]
  end

  def part_1({:down, dist}, x: x, y: y) do
    [x: x, y: y + dist]
  end

  def part_1({:up, dist}, x: x, y: y) do
    [x: x, y: y - dist]
  end

  def part_2(commands) when is_list(commands) do
    commands
    |> Enum.reduce([x: 0, y: 0, aim: 0], &part_2/2)
    |> calculate_answer()
  end

  def part_2({:forward, dist}, x: x, y: y, aim: aim) do
    [x: x + dist, y: y + aim * dist, aim: aim]
  end

  def part_2({:down, dist}, x: x, y: y, aim: aim) do
    [x: x, y: y, aim: aim + dist]
  end

  def part_2({:up, dist}, x: x, y: y, aim: aim) do
    [x: x, y: y, aim: aim - dist]
  end

  def calculate_answer(x: x, y: y), do: x * y
  def calculate_answer(x: x, y: y, aim: _), do: x * y
end
